hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/tagdetails/AbstractTagToken.java

```diff
@@ -110,267 +91,96 @@ protected void onBaseEntityEvent(final BaseUIEntityEvent<T> baseEntityEvent) {
-            button.setCaptionAsHtml(true);
-        }
-
-        private void onTokenSearch(final Object tokenId) {
-            assignTag(getItemNameProperty(tokenId).getValue().toString());
-            removeTagAssignedFromCombo((Long) tokenId);
-        }
-
-        private void tokenClick(final Object tokenId) {
-            final Item item = tokenField.getContainerDataSource().addItem(tokenId);
-            item.getItemProperty(NAME_PROPERTY).setValue(tagDetails.get(tokenId).getName());
-            item.getItemProperty(COLOR_PROPERTY).setValue(tagDetails.get(tokenId).getColor());
-            unassignTag(tagDetails.get(tokenId).getName());
+    @Override
+    public void assignTag(final String tagName) {
+        final TagData tagData = tagDetailsByName.get(tagName);
+        if (tagData != null) {
+            assignTag(tagData);
         }
-
-        private Property getItemNameProperty(final Object tokenId) {
-            final Item item = tokenField.getContainerDataSource().getItem(tokenId);
-            return item.getItemProperty(NAME_PROPERTY);
-        }
-
     }
 
-    private static String getColor(final Item item) {
-        if (item.getItemProperty(COLOR_PROPERTY).getValue() != null) {
-            return (String) item.getItemProperty(COLOR_PROPERTY).getValue();
-        } else {
-            return SPUIDefinitions.DEFAULT_COLOR;
+    @Override
+    public void unassignTag(final String tagName) {
+        final TagData tagData = tagDetailsByName.get(tagName);
+        if (tagData != null) {
+            unassignTag(tagData);
         }
     }
 
-    private static String getTagName(final Item item) {
-        return (String) item.getItemProperty(NAME_PROPERTY).getValue();
-    }
+    protected abstract void assignTag(TagData tagData);
 
-    protected void removePreviouslyAddedTokens() {
-        tokensAdded.keySet().forEach(previouslyAddedToken -> tokenField.removeToken(previouslyAddedToken));
-    }
+    protected abstract void unassignTag(TagData tagData);
 
     protected Long getTagIdByTagName(final Long tagId) {
-        return tagDetails.entrySet().stream().filter(entry -> entry.getValue().getId().equals(tagId)).findAny()
-                .map(entry -> entry.getKey()).orElse(null);
-
-    }
-
-    protected void removeTokenItem(final Long tokenId, final String name) {
-        tokenField.removeToken(tokenId);
-        tagDetails.remove(tokenId);
-        setContainerPropertValues(tokenId, name, tokensAdded.get(tokenId).getColor());
+        return tagDetailsById.entrySet().stream().filter(entry -> entry.getValue().getId().equals(tagId)).findAny()
+                .map(Entry::getKey).orElse(null);
     }
 
-    protected void removeTagFromCombo(final Long deletedTagId) {
-        if (deletedTagId != null) {
-            container.removeItem(deletedTagId);
+    protected boolean checkAssignmentResult(final List<? extends NamedEntity> assignedEntities,
```

Comment:
use `Identifiable` instead of `NamedEntity `, you are not interested in Name here